server/util/redact/redact.go

```diff
@@ -62,19 +74,70 @@ func stripURLSecrets(input string) string {
 	return urlSecretRegex.ReplaceAllString(input, "")
 }
 
+// Strips URL secrets from the provided flag value, if there is a value.
+func stripUrlSecretsFromFlag(input string) string {
+	ck, cv := splitCombinedForm(input)
+	if cv == "" {
+		// No flag value to redact; keep flag name as-is.
+		return input
+	}
+	return ck + "=" + stripURLSecrets(cv)
+}
+
+// Strips URL secrets from all of the provided multi-flags. These are flags
+// that are specified as:
+//
+//	--flag_name=ONE=foo,TWO=BAR,THREE=baz
+//
+// This function will strip secrets from the "foo", "bar", and "baz" components,
+// leaving "--flag_name", "ONE", "TWO", and "THREE" intact.
+func stripUrlSecretsFromMultiFlag(input string) string {
+	key, val := splitCombinedForm(input)
+	return key + "=" + stripUrlSecretsFromMultiFlagValue(val)
+}
+
+// Same as above but expects to not get the "--flag_name=" part.
+func stripUrlSecretsFromMultiFlagValue(input string) string {
+	subFlags := splitMultiFlag(input)
+	for i, flag := range subFlags {
+		subFlags[i] = stripUrlSecretsFromFlag(flag)
+	}
+	return strings.Join(subFlags, ",")
+}
+
+// Splits the provided multi-flags into an array of regular ole' flags. So,
+// "ONE=foo,TWO=bar,THREE=baz" becomes ["ONE=foo", "TWO=bar", "THREE=baz"]
+func splitMultiFlag(input string) []string {
+	output := make([]string, strings.Count(input, "="))
+	subFlags := multiFlagKeyRegex.FindAllStringIndex(input, -1 /* return all matches */)
+	for i := 0; i < len(subFlags); i++ {
+		if i+1 >= len(subFlags) {
+			output[i] = input[subFlags[i][0]:len(input)]
+		} else {
+			output[i] = strings.TrimRight(input[subFlags[i][0]:subFlags[i+1][0]], ",")
+		}
+	}
+	return output
+}
+
+func isKnownMultiFlag(input string) bool {
+	for _, knownMultiFlag := range knownMultiFlags {
+		if strings.HasPrefix(input, knownMultiFlag) || strings.HasPrefix(input, "--"+knownMultiFlag) {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
Would it be practical to treat non-multi flags as a special case of multi flags with a length of 1, so that everything uses the same multi-flag logic? Hoping we can avoid introducing separate special handling for single- and multi- valued flags.

Reply:
I thought about that. The drawback would be that if a single-value flag has an equals sign in it, then the portion before the equals sign wouldn't be subject to redaction, so I figured it'd be better to be cautious and only do this for a set of known flags. Maybe that's not an issue, though? WDYT?

Reply:
Ah fair point. I'm cool with leaving it as-is :+1: